### PR TITLE
[Refactor] Simplify MissingClientsRequestStrategy

### PR DESCRIPTION
**Pull request by typfel**
## What's new in this PR?

Simplify `MissingClientsRequestStrategy` and make it using type safe structs for decoding and encoding payloads. This is done  in preparation to support the federation aware `/users/list-prekeys` in addition to the existing legacy `/users/prekeys` endpoint.

### CODE REVIEW

File: Sources/Helpers/Dictionary+TakeFirst.swift [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        guard keys.count > count else {
+            return self
+        }
+
+        return reduce(into: Self()) { (result, element) in
+            let (key, value) = element
+            if result.keys.count < count {
+                result[key] = value
+            }
+        }
```
[the remark this answers is not in the document]
Comment (typfel) [reply Sources/Helpers/Dictionary+TakeFirst.swift:37]: Deleted this extension in part 2

File: Sources/Request Strategies/User Clients/MissingClientsRequestFactory.swift [hunk 2]
```diff
@@ -19,54 +19,48 @@
...
+        return self.reduce(into: initial) { (result, client) in
+            guard let userID = client.user?.remoteIdentifier.transportString(),
+                  let clientID = client.remoteIdentifier
+            else {
+                return
+            }
+
+            var clientList = result[userID] ?? []
+            clientList.append(clientID)
+            result[userID] = clientList
```
[the remark this answers is not in the document]
Comment (typfel) [reply Sources/Request Strategies/User Clients/MissingClientsRequestFactory.swift:38]: Nice!

File: Sources/Payloads/Payload+Processing.swift [hunk 3]
```diff
@@ -59,3 +59,68 @@ extension Array where Array.Element == Payload.UserClient {
...
+                                                                          selfClient: selfClient)
+                } else {
+                    missingClient.markClientAsInvalidAfterFailingToRetrievePrekey(selfClient: selfClient)
+                }
+
+
+            }
+        }
+
+        return (selfClient.missingClients?.count ?? 0) > 0
```
[the remark this answers is not in the document]
Comment (typfel) [reply Sources/Payloads/Payload+Processing.swift:89]: Exactly, this is part of the upstream modified sync protocol. I'll add a comment.

File: Sources/Request Strategies/User Clients/MissingClientsRequestFactory.swift [hunk 4]
```diff
@@ -19,54 +19,46 @@
...
         self.pageSize = pageSize
     }
-    
-    public func fetchMissingClientKeysRequest(_ missingClients: Set<UserClient>) -> ZMUpstreamRequest! {
-        let map = MissingClientsMap(Array(missingClients), pageSize: pageSize)
-        let request = ZMTransportRequest(path: "/users/prekeys", method: ZMTransportRequestMethod.methodPOST, payload: map.payload as ZMTransportData?)
-        return ZMUpstreamRequest(keys: Set(arrayLiteral: ZMUserClientMissingKey), transportRequest: request, userInfo: map.userInfo)
+
+    public func fetchPrekeys(for missingClients: Set<UserClient>) -> ZMUpstreamRequest {
+        let payloadData = missingClients.clientListByUserID.takeFirst(pageSize).payloadData()!
```
Comment (KaterinaWire) [inline Sources/Request Strategies/User Clients/MissingClientsRequestFactory.swift:50]: Is it safe to force unwrap here?

Comment (typfel) [reply Sources/Request Strategies/User Clients/MissingClientsRequestFactory.swift:50]: No, good catch! I'll remove this force unwrap

**Review by johnxnguyen — APPROVED** [review]

File: Sources/Request Strategies/User Clients/MissingClientsRequestFactory.swift (hunk 4 above)
Comment (KaterinaWire) [reply Sources/Request Strategies/User Clients/MissingClientsRequestFactory.swift:50]: I saw the same method in another PR, and there's no force unwrap in that PR. Are you going to merge that PR and not fix it here?

Comment (typfel) [reply Sources/Request Strategies/User Clients/MissingClientsRequestFactory.swift:50]: Yeah, I did the change on the other PR. I'll merge them into one.

**Review by KaterinaWire — APPROVED** [review]

**Pull request merged**